packages/odata-service-inquirer/src/prompts/datasources/sap-system/external-services/value-help-download.ts

```diff
@@ -0,0 +1,168 @@
+ *
+ * @param connectionValidator - connection validator instance
+ * @param promptNamespace - prompt namespace
+ * @param convertedMetadataRef - converted metadata reference
+ * @param convertedMetadataRef.convertedMetadata - converted metadata
+ * @returns value help download prompt
+ */
+export function getValueHelpDownloadPrompt(
+    connectionValidator: ConnectionValidator,
+    promptNamespace?: string,
+    convertedMetadataRef?: {
+        convertedMetadata: ConvertedMetadata | undefined;
+    }
+): ConfirmQuestion {
+    const promptNamespacePart = `${promptNamespace ? promptNamespace + ':' : ''}`;
+    const promptName = `${promptNamespacePart}${promptNames.valueHelpDownload}`;
+    const servicePromptName = `${promptNamespacePart}${promptNames.serviceSelection}`;
+
+    let externalServiceRefs: ExternalServiceReference[];
+    let externalServiceMetadata: ExternalService[];
+
+    const question: ConfirmQuestion = {
+        when: (answers: Record<string, any> | undefined) => {
+            const service = answers?.[servicePromptName] as ServiceAnswer | undefined;
+            const servicePath = service?.servicePath;
+            if (servicePath && convertedMetadataRef?.convertedMetadata) {
+                externalServiceRefs = getExternalServiceReferences(
+                    servicePath,
+                    convertedMetadataRef?.convertedMetadata
+                );
+                return externalServiceRefs.length > 0;
+            }
+            return false;
+        },
+        name: promptName,
+        type: 'confirm',
+        message: t('prompts.valueHelpDownload.message'),
+        validate: async (downloadMetadata: boolean) => {
+            delete PromptState.odataService.valueListMetadata;
+
+            // Send telemetry when prompt is answered
+            const telemetryData = createValueHelpTelemetryData({
+                userChoseToDownload: downloadMetadata
+            });
+            sendTelemetryEvent(telemEventValueHelpDownloadPrompted, telemetryData.properties);
+
+            if (downloadMetadata && connectionValidator.serviceProvider instanceof AbapServiceProvider) {
+                const startTime = Date.now();
+                try {
+                    externalServiceMetadata = await (
+                        connectionValidator.serviceProvider as AbapServiceProvider
+                    ).fetchExternalServices(externalServiceRefs);
+                    const downloadTimeMs = Date.now() - startTime;
+
+                    if (externalServiceMetadata.length > 0) {
+                        PromptState.odataService.valueListMetadata = externalServiceMetadata;
+                    } else {
+                        LoggerHelper.logger.info(t('warnings.noExternalServiceMetdataFetched'));
+                    }
+
+                    // Send telemetry with measurements for numeric metrics
+                    const telemetryData = createValueHelpTelemetryData({
```

Comment:
Could we move all this telemetry code into another function and re-use for the three telemetry cases?

Reply:
Thanks, I have updated